### PR TITLE
dpdk: EnsureDPDKBuffer

### PR DESCRIPTION
**Pull request by tbarbette**
Ensure that a packet is backed by a DPDK buffer. Allows to preload
packets and send them faster as there will be no copy when sending
the packet to DPDK.

### CODE REVIEW

File: lib/packet.cc [hunk 1]
```diff
@@ -613,6 +613,21 @@ Packet::make(unsigned char *data, uint32_t length,
...
 
+/** @brief Copy the content and annotations of another packet (userlevel).
+ * @param source packet
+ * @param headroom for the new packet
+ */
+void
+Packet::copy(Packet* p, int headroom)
+{
+	_data = _head + headroom;
+	memcpy(_data,p->data(),p->length());
```
Comment (kohler) [inline lib/packet.cc]: urgh no this needs at the very least a bunch of length checks.

File: lib/packet.cc [hunk 2]
```diff
@@ -613,6 +613,21 @@ Packet::make(unsigned char *data, uint32_t length,
...
+	_tail = _data + p->length();
+	copy_annotations(p);
+    ptrdiff_t shift = _data - p->_data;
```
Comment (kohler) [inline lib/packet.cc]: ...spacing

File: lib/packet.cc [hunk 3]
```diff
@@ -613,6 +613,21 @@ Packet::make(unsigned char *data, uint32_t length,
...
+    set_mac_header(p->mac_header() ? p->mac_header() + shift : 0);
+    set_network_header(p->network_header() ? p->network_header() + shift : 0, p->network_header_length());
```
Comment (kohler) [inline lib/packet.cc]: think this is wrong—it's setting OUR mac header to a pointer located in a DIFFERENT packet; want a computation involving _data

Comment (tbarbette) [inline lib/packet.cc]: I think you missed that shift is the difference between _data and p->_data, so it should be good.

File: lib/packet.cc (hunk 1 above)
Comment (tbarbette) [inline lib/packet.cc]: Added a check for the total length check with the new header. Other idea?

File: lib/packet.cc (hunk 3 above)
Comment (tbarbette) [inline lib/packet.cc]: Better like that?
Thanks for the input.

**Comment by kohler** [comment]
Thanks!

**Pull request merged**